### ISSUE TITLE
Mixed visibility of property getter/setter making property setter not usable in method / constructor invocation.

### ISSUE THREAD

**Issue by smoothdeveloper** [dotnet/fsharp#8351]
```csharp
namespace csharp
{
    public class Class
    {
        public int Prop { set; private get; }
    }
}
```

```fsharp
type Maker =
  static member mkCs () = csharp.Class()
let a = csharp.Class(Prop=1)
let aa = Maker.mkCs(Prop=1)
```

I'd expect that F# code to work, it works with VB.net and F# defined classes, but not C#.

edit: it doesn't work whenever there is private getter from any non F# IL.

**Comment by smoothdeveloper** [comment]
The IL for reference

C#:
```csharp
.class public auto ansi beforefieldinit csharp.Class
	extends [mscorlib]System.Object
{
	// Fields
	.field private int32 '<Prop>k__BackingField'
	.custom instance void [mscorlib]System.Runtime.CompilerServices.CompilerGeneratedAttribute::.ctor() = (
		01 00 00 00
	)
	.custom instance void [mscorlib]System.Diagnostics.DebuggerBrowsableAttribute::.ctor(valuetype [mscorlib]System.Diagnostics.DebuggerBrowsableState) = (
		01 00 00 00 00 00 00 00
	)

	// Methods
	.method public hidebysig specialname 
		instance void set_Prop (
			int32 'value'
		) cil managed 
	{
		.custom instance void [mscorlib]System.Runtime.CompilerServices.CompilerGeneratedAttribute::.ctor() = (
			01 00 00 00
		)
		// Method begins at RVA 0x2050
		// Code size 8 (0x8)
		.maxstack 8

		IL_0000: ldarg.0
		IL_0001: ldarg.1
		IL_0002: stfld int32 csharp.Class::'<Prop>k__BackingField'
		IL_0007: ret
	} // end of method Class::set_Prop

	.method private hidebysig specialname 
		instance int32 get_Prop () cil managed 
	{
		.custom instance void [mscorlib]System.Runtime.CompilerServices.CompilerGeneratedAttribute::.ctor() = (
			01 00 00 00
		)
		// Method begins at RVA 0x2059
		// Code size 7 (0x7)
		.maxstack 8

		IL_0000: ldarg.0
		IL_0001: ldfld int32 csharp.Class::'<Prop>k__BackingField'
		IL_0006: ret
	} // end of method Class::get_Prop

	.method public hidebysig specialname rtspecialname 
		instance void .ctor () cil managed 
	{
		// Method begins at RVA 0x2061
		// Code size 8 (0x8)
		.maxstack 8

		IL_0000: ldarg.0
		IL_0001: call instance void [mscorlib]System.Object::.ctor()
		IL_0006: nop
		IL_0007: ret
	} // end of method Class::.ctor

	// Properties
	.property instance int32 Prop()
	{
		.get instance int32 csharp.Class::get_Prop()
		.set instance void csharp.Class::set_Prop(int32)
	}

} // end of class csharp.Class
```

VB:
```csharp
.class public auto ansi basic.BasicClass
	extends [mscorlib]System.Object
{
	// Fields
	.field private int32 v

	// Methods
	.method public specialname rtspecialname 
		instance void .ctor () cil managed 
	{
		// Method begins at RVA 0x20fc
		// Code size 7 (0x7)
		.maxstack 8

		IL_0000: ldarg.0
		IL_0001: call instance void [mscorlib]System.Object::.ctor()
		IL_0006: ret
	} // end of method BasicClass::.ctor

	.method public specialname 
		instance void set_Prop (
			int32 'value'
		) cil managed 
	{
		// Method begins at RVA 0x2104
		// Code size 9 (0x9)
		.maxstack 8

		IL_0000: nop
		IL_0001: ldarg.0
		IL_0002: ldarg.1
		IL_0003: stfld int32 basic.BasicClass::v
		IL_0008: ret
	} // end of method BasicClass::set_Prop

	// Properties
	.property instance int32 Prop()
	{
		.set instance void basic.BasicClass::set_Prop(int32)
	}

} // end of class basic.BasicClass

```

**Comment by smoothdeveloper** [comment]
Culprit:
https://github.com/dotnet/fsharp/blob/70883fb00a867f6e81aa9c7cecdd9212c4e78d93/src/fsharp/AccessibilityLogic.fs#L247-L256

Top level callsite of interest:

https://github.com/dotnet/fsharp/blob/70883fb00a867f6e81aa9c7cecdd9212c4e78d93/src/fsharp/MethodCalls.fs#L327

In vb case, there is no getter (not sure it supports declaring public setter with private getter) so it goes in the second part of the match, in C# case, it stops at the getter.

I'm thinking making another method similar to `GetILAccessOfILPropInfo` but where it is the setter only which is checked would do the trick of solving this issue.

**Comment by abelbraaksma** [comment]
Just wondering, it used to be possible to call the `set_Prop` directly from F#. If that's still possible, that could be used as workaround perhaps, until this is fixed?

**Comment by smoothdeveloper** [comment]
@abelbraaksma this is indeed possible, even with the usual `x.Prop <- value`; I was specifically interested in the setter as method named argument case.

regarding my initial reasoning for a fix:

> I'm thinking making another method similar to GetILAccessOfILPropInfo but where it is the setter only which is checked would do the trick of solving this issue.

This won't fly, the whole property access check logic is based on assumption that getter visibility trumps the setter one which isn't even looked at if getter exists, the F# code compile the getter as `internal`, so putting it in a library should also fail.

https://github.com/dotnet/fsharp/blob/792dd92dff2d4b7a8ab7d7a60aa7a16bb39e9939/tests/fsharp/core/members/set-only-property/fs.fs#L3-L4

I'm updating the test in #8352 to highlight that as well.

The spec doesn't mention anything about visibility or getter § 14.4.4.e:

> If the return type of M, before the application of any type arguments ActualTypeArgs,
contains a settable property name,then name is the target. The available properties
include any property extension members of type, found by consulting the
ExtensionsInScope table.

**Comment by smoothdeveloper** [comment]
Renamed this issue more apropriately now that the root cause is identified.

**Comment by cartermp** [comment]
Resolved in #8351

**Comment by abelbraaksma** [comment]
> Resolved in #8351  

Great! But I think you mean #8354? 

And if it's fixed, why is it on the backlog? I thought that was for stuff that wasn't fixed yet.

**Comment by cartermp** [comment]
Yes, 8354, my bad.

Backlog is a bucket. Things get re-assigned as releases occur.
